### ISSUE TITLE
sort is slow on 0.6 and falls off a cliff at weird sizes

### ISSUE THREAD

**Issue by chethega** [Jutho/TupleTools.jl#3]
Very cool package!

Unfortunately the sort falls off a cliff once the tuples / types are a little larger (on 0.7, it is always slow on 0.6). That's OK for me but a little weird, and the prospect of having to adjust my integer sizes is slightly annoying (not shown: unsigned integers fall off the cliff earlier, it appears like the largest good sort is `NTuple{13,Int32}` ).

```
julia> for n=4:12
       for T in (Int16,Int32,Int64)
       println("sorting Tuple with $(n) elements of type $(T)")
       t=(collect(T(1):T(n))...,)
       @btime TupleTools.sort($t)
       end
       end;

sorting Tuple with 4 elements of type Int16
  22.250 ns (0 allocations: 0 bytes)
sorting Tuple with 4 elements of type Int32
  28.794 ns (0 allocations: 0 bytes)
sorting Tuple with 4 elements of type Int64
  27.489 ns (0 allocations: 0 bytes)
sorting Tuple with 5 elements of type Int16
  46.665 ns (0 allocations: 0 bytes)
sorting Tuple with 5 elements of type Int32
  41.435 ns (0 allocations: 0 bytes)
sorting Tuple with 5 elements of type Int64
  545.319 ns (3 allocations: 112 bytes)
sorting Tuple with 6 elements of type Int16
  59.301 ns (0 allocations: 0 bytes)
sorting Tuple with 6 elements of type Int32
  56.688 ns (0 allocations: 0 bytes)
sorting Tuple with 6 elements of type Int64
  1.089 ��s (6 allocations: 240 bytes)
sorting Tuple with 7 elements of type Int16
  74.307 ns (0 allocations: 0 bytes)
sorting Tuple with 7 elements of type Int32
  71.506 ns (0 allocations: 0 bytes)
sorting Tuple with 7 elements of type Int64
  1.696 ��s (9 allocations: 368 bytes)
sorting Tuple with 8 elements of type Int16
  90.244 ns (0 allocations: 0 bytes)
sorting Tuple with 8 elements of type Int32
  87.202 ns (0 allocations: 0 bytes)
sorting Tuple with 8 elements of type Int64
  2.378 ��s (12 allocations: 528 bytes)
sorting Tuple with 9 elements of type Int16
  107.682 ns (0 allocations: 0 bytes)
sorting Tuple with 9 elements of type Int32
  106.816 ns (0 allocations: 0 bytes)
sorting Tuple with 9 elements of type Int64
  3.126 ��s (15 allocations: 688 bytes)
sorting Tuple with 10 elements of type Int16
  125.118 ns (0 allocations: 0 bytes)
sorting Tuple with 10 elements of type Int32
  126.416 ns (0 allocations: 0 bytes)
sorting Tuple with 10 elements of type Int64
  3.861 ��s (18 allocations: 880 bytes)
sorting Tuple with 11 elements of type Int16
  145.642 ns (0 allocations: 0 bytes)
sorting Tuple with 11 elements of type Int32
  152.863 ns (0 allocations: 0 bytes)
sorting Tuple with 11 elements of type Int64
  4.711 ��s (20 allocations: 992 bytes)
sorting Tuple with 12 elements of type Int16
  998.091 ns (3 allocations: 80 bytes)
sorting Tuple with 12 elements of type Int32
  180.781 ns (0 allocations: 0 bytes)
sorting Tuple with 12 elements of type Int64
  5.573 ��s (22 allocations: 1.09 KiB)
```

**Comment by Jutho** [comment]
I get strangely different results, in particular, I always get 2 allocations wth copy pasting your code, both in v0.6.3 and in master.

Note that there is an (undocumented) `_sort` method, which does not use keyword arguments (this dates back from before keyword arguments where fast) and seems to perform well in all of these cases (just replace to `TupleTools._sort`).

However, since keyword arguments should no longer be a problem, I'll have to check what is going on.

**Comment by Jutho** [comment]
This should be fixed now, both in julia 0.6 and 0.7, with the latest master (which I also tagged as a version).

**Comment by Jutho** [comment]
I.e. I obtain this:
```
sorting Tuple with 4 elements of type Int16
  17.997 ns (0 allocations: 0 bytes)
sorting Tuple with 4 elements of type Int32
  17.482 ns (0 allocations: 0 bytes)
sorting Tuple with 4 elements of type Int64
  13.286 ns (0 allocations: 0 bytes)
sorting Tuple with 5 elements of type Int16
  27.229 ns (0 allocations: 0 bytes)
sorting Tuple with 5 elements of type Int32
  22.421 ns (0 allocations: 0 bytes)
sorting Tuple with 5 elements of type Int64
  17.712 ns (0 allocations: 0 bytes)
sorting Tuple with 6 elements of type Int16
  32.107 ns (0 allocations: 0 bytes)
sorting Tuple with 6 elements of type Int32
  27.440 ns (0 allocations: 0 bytes)
sorting Tuple with 6 elements of type Int64
  23.225 ns (0 allocations: 0 bytes)
sorting Tuple with 7 elements of type Int16
  38.935 ns (0 allocations: 0 bytes)
sorting Tuple with 7 elements of type Int32
  30.380 ns (0 allocations: 0 bytes)
sorting Tuple with 7 elements of type Int64
  26.943 ns (0 allocations: 0 bytes)
sorting Tuple with 8 elements of type Int16
  41.843 ns (0 allocations: 0 bytes)
sorting Tuple with 8 elements of type Int32
  36.980 ns (0 allocations: 0 bytes)
sorting Tuple with 8 elements of type Int64
  36.896 ns (0 allocations: 0 bytes)
sorting Tuple with 9 elements of type Int16
  48.413 ns (0 allocations: 0 bytes)
sorting Tuple with 9 elements of type Int32
  44.351 ns (0 allocations: 0 bytes)
sorting Tuple with 9 elements of type Int64
  40.845 ns (0 allocations: 0 bytes)
sorting Tuple with 10 elements of type Int16
  56.738 ns (0 allocations: 0 bytes)
sorting Tuple with 10 elements of type Int32
  51.020 ns (0 allocations: 0 bytes)
sorting Tuple with 10 elements of type Int64
  48.784 ns (0 allocations: 0 bytes)
sorting Tuple with 11 elements of type Int16
  61.623 ns (0 allocations: 0 bytes)
sorting Tuple with 11 elements of type Int32
  60.491 ns (0 allocations: 0 bytes)
sorting Tuple with 11 elements of type Int64
  59.068 ns (0 allocations: 0 bytes)
sorting Tuple with 12 elements of type Int16
  70.543 ns (0 allocations: 0 bytes)
sorting Tuple with 12 elements of type Int32
  64.976 ns (0 allocations: 0 bytes)
sorting Tuple with 12 elements of type Int64
  65.102 ns (0 allocations: 0 bytes)
```
